docs/spec/overload.rst

```diff
@@ -99,8 +109,349 @@ variable is not sufficient.
+
+
+Example 2::
+
+  @overload
+  def example2(x: int, y: str, z: int) -> str: ...
+  @overload
+  def example2(x: int, y: int, z: int) -> int: ...
+
+  def test(values: list[str | int]):
+      # In this example, argument type expansion is
+      # performed on the first two arguments. Expansion
+      # of the third is unnecessary.
+      r1 = example2(1, values[0], 1)
+      reveal_type(r1)  # Should reveal str | int
+
+      # Here, the types of all three arguments are expanded
+      # without success.
+      example2(values[0], values[1], values[2])  # Error in step 3
+
+
+Example 3::
+
+  @overload
+  def example3(x: int, /) -> tuple[int]: ...
+  @overload
+  def example3(x: int, y: int, /) -> tuple[int, int]: ...
+  @overload
+  def example3(*args: int) -> tuple[int, ...]: ...
+
+  def test(val: list[int]):
+      # Step 1 eliminates second overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the first
+      # overload.
+      r1 = example3(1)
+      reveal_type(r1)  # Should reveal tuple[int]
+
+      # Step 1 eliminates first overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the second
+      # overload.
+      r2 = example3(1, 2)
+      reveal_type(r2)  # Should reveal tuple[int, int]
+
+      # Step 1 doesn't eliminate any overloads. Step 4
+      # picks the third overload.
+      r3 = example3(*val)
+      reveal_type(r3)  # Should reveal tuple[int, ...]
+
+
+Example 4::
+
+  @overload
+  def example4(x: list[int], y: int) -> int: ...
+  @overload
+  def example4(x: list[str], y: str) -> int: ...
+  @overload
+  def example4(x: int, y: int) -> list[int]: ...
+
+  def test(v1: list[Any], v2: Any):
+      # Step 2 eliminates the third overload. Step 5
+      # determines that first and second overloads
+      # both apply and are ambiguous due to Any, but
+      # return types are consistent.
+      r1 = example4(v1, v2)
+      reveal_type(r1)  # Reveals int
+
+      # Step 2 eliminates the second overload. Step 5
+      # determines that first and third overloads
+      # both apply and are ambiguous due to Any, and
+      # the return types are inconsistent.
+      r2 = example4(v2, 1)
+      reveal_type(r2)  # Should reveal Any
+
+
+.. _argument-type-expansion:
+
+Argument type expansion
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When performing argument type expansion, a type that is equivalent to
+a union of a finite set of subtypes should be expanded into its constituent
+subtypes. This includes the following cases.
+
+1. Explicit unions: Each subtype of the union should be considered as a
+separate argument type. For example, the type ``int | str`` should be expanded
+into ``int`` and ``str``.
+
+2. ``bool`` should be expanded into ``Literal[True]`` and ``Literal[False]``.
+
+3. ``Enum`` types (other than those that derive from ``enum.Flag``) should
+be expanded into their literal members.
+
+4. ``type[A | B]`` should be expanded into ``type[A]`` and ``type[B]``.
```

Comment:
This might be more clearly expressed by simply saying that `type[A | B]` is equivalent to `type[A] | type[B]`, but that's for a different chapter of the spec.

Reply:
The `type` section already specifies that `type[A | B]` is equivalent to `type[A] | type[B]`. All of the expansion rules in this list are examples of type equivalencies.